engine/src/silo/mod.rs

```diff
@@ -0,0 +1,196 @@
+use aurora_engine_sdk::io::IO;
+#[cfg(feature = "contract")]
+use aurora_engine_sdk::{env::Env, types::SdkUnwrap};
+use aurora_engine_types::account_id::AccountId;
+use aurora_engine_types::storage::{bytes_to_key, KeyPrefix};
+use aurora_engine_types::types::{Address, Wei};
+use aurora_engine_types::AsBytes;
+
+#[cfg(feature = "contract")]
+use crate::engine::EngineErrorKind;
+use crate::prelude::Vec;
+
+use parameters::{WhitelistArgs, WhitelistKindArgs, WhitelistStatusArgs};
+use whitelist::Whitelist;
+pub use whitelist::WhitelistKind;
+
+pub mod parameters;
+mod whitelist;
+
+const GAS_COST_KEY: &[u8] = b"GAS_COST_KEY";
+
+/// Return fixed gas cost.
+pub fn get_fixed_gas_cost<I: IO>(io: &I) -> Option<Wei> {
+    let key = fixed_gas_cost_key();
+    io.read_u256(&key).ok().map(Wei::new)
+}
+
+/// Set fixed gas cost.
+pub fn set_fixed_gas_cost<I: IO>(io: &mut I, cost: Option<Wei>) {
+    let key = fixed_gas_cost_key();
+
+    if let Some(cost) = cost {
+        io.write_storage(&key, &cost.to_bytes());
+    } else {
+        io.remove_storage(&key);
+    }
+}
+
+/// Add an entry to a white list depending on a kind of list types in provided arguments.
+pub fn add_entity_to_whitelist<I: IO + Copy>(io: &I, args: &WhitelistArgs) {
+    let (kind, entry) = get_kind_and_entry(args);
+    Whitelist::init(io, kind).add(entry);
+}
+
+/// Add an entries to a white list depending on a kind of list types in provided arguments.
+pub fn add_entry_to_whitelist_batch<I: IO + Copy, A: IntoIterator<Item = WhitelistArgs>>(
+    io: &I,
+    entries: A,
+) {
+    for entry in entries {
+        add_entity_to_whitelist(io, &entry);
+    }
+}
+
+/// Remove an entries to a white list depending on a kind of list types in provided arguments.
+pub fn remove_entry_from_whitelist<I: IO + Copy>(io: &I, args: &WhitelistArgs) {
+    let (kind, entry) = get_kind_and_entry(args);
+    Whitelist::init(io, kind).remove(entry);
+}
+
+/// Set status of the provided white list.
+pub fn set_whitelist_status<I: IO + Copy>(io: &I, args: &WhitelistStatusArgs) {
+    whitelist::set_whitelist_status(io, args);
+}
+
+/// Return status of the provided white list.
+pub fn get_whitelist_status<I: IO + Copy>(io: &I, args: &WhitelistKindArgs) -> WhitelistStatusArgs {
+    whitelist::get_whitelist_status(io, args)
+}
+
+/// Check if the calling user is admin or owner.
+#[cfg(feature = "contract")]
+pub fn assert_admin<I: IO + Env + Copy>(io: &I) -> Result<(), EngineErrorKind> {
+    if !is_owner(io) && !is_admin(io) {
+        return Err(EngineErrorKind::NotAllowed);
+    }
+
+    Ok(())
+}
+
+/// Check if user has writes to deploy EVM code.
+pub fn is_allow_deploy<I: IO + Copy>(io: &I, account: &AccountId, address: &Address) -> bool {
+    let admin_list = Whitelist::init(io, WhitelistKind::Admin);
```

Comment:
@aleksuss this part can be simplified a bit if you reuse `is_admin` function here? Optionally, you can also define a similar `is_evm_admin` fn.